File: cli/command.py
import sys
import logging
import os
import threading

class Command:
	def __init__(self, kernel):
		logging.info("CLI:\tcommand init.")
		self._kernel = kernel

	def update(self):
		self._kernel.update()

	def halt(self):
		self._kernel.halt()

	def throttle(self):
		var = raw_input("How much throttle [0-100]: ").strip()
		self._kernel.set_throttle(var)

	def steering(self):
		var = raw_input("which angle [-180-180]: ").strip()
		self._kernel.set_steering(var)

	def enable_nav(self):
		#enable the automatic navigation of the boat
		logging.info("CLI:\tnav started.")
		threading.Thread(target=self._kernel.run_navigation).start()
	
	def disable_nav(self):
		logging.info("CLI:\tnav stopping.")
		self._kernel._navigationCanRun = False #fuck threading joins
	
	def add_point(self):
		lat = raw_input("Latitude: ").strip()
		lon = raw_input("Longitude: ").strip()
		self._kernel.add_point(lat, lon)
		logging.info("CLI:\tPoint added.")

	def thread_info(self):
		print "Active Threads : " + str(threading.activeCount())

def help():
	#we want it to look nice hu?
	print "-------------------|>---------------"
	print "-------------------|----------------"
	print "~~~~~~~~~~~~\INVENAVI/~~~~~~~~~~~~~~"
	print "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
	print "update     : reads all the sensors"
	print "halt       : stop all motors/servo's"
	print "throttle   : set motor"
	print "steering   : set servo angle"
	print "add point  : add waypoint for nav"
	print "enablenav  : enable navigation"
	print "disablenav : disable navigation"
	print "threadinfo : show info about threads"
	print "takepicture: take a picture"
	print "____________________________________"
	print "to exit press ctrl+c"
	print ""

# This will be to handle input for a function we don't have
def fail():
	print "You gave a bad function name , type help!"

def run(kernel):
	comd = Command(kernel)
	#dictonary
	funcs = {"threadinfo" : comd.thread_info, "addpoint" : comd.add_point,  "update": comd.update, "throttle": comd.throttle, "steering": comd.steering, "halt": comd.halt, "enablenav": comd.enable_nav, "disablenav": comd.disable_nav, "help": help}

	while 1:
		try:
			var = raw_input("Command: ").strip()
			funcs.get(var, fail)()
		except KeyboardInterrupt:
			break

File: command.js
// Constructor
var Command = function (kernel) {
  // always initialize all instance properties
  this.kernel = kernel;
  this.stdin = process.stdin;
  this.stdout = process.stdout;

  this.start = function() {
    console.log('starting cli mode');
    this.stdin.resume();
    this.stdin.setEncoding( 'utf8' );
    this.complete();

    // on any data into stdin
    this.stdin.on( 'data', function( data ){
      data = data.toString().trim();

      if (data == 'help') {
        this.help();
      } else if (data == 'start') {
        this.startComponents();
      } else if (data == 'stop') {
        this.stopComponents();
      } else if (data == 'steering') {
        this.steering();
      } else if (data == 'throttle') {
        this.throttle();
      } else if (data == 'update') {
        this.update();
      } else if (data == 'config') {
        this.config();
      } else if (data == 'halt') {
        this.halt();
      } else if ( data == '\u0003' || data == 'quit' || data == 'exit') {
        this.exit();
      } else {
        this.complete();
      }

    }.bind(this));

    return this;
  };

  this.complete = function() {
    this.stdout.write('invenavi$ ');
  }

  this.help = function() {
      console.log("-------------------|>---------------");
      console.log("-------------------|----------------");
      console.log("~~~~~~~~~~~~\\INVENAVI/~~~~~~~~~~~~~~");
      console.log("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
      console.log("update     : reads all the sensors");
      console.log("config     : dump config");
      console.log("reset      : reset all engines and servo's");
      console.log("halt       : stop all motors/servo's");
      console.log("start      : start all sensors");
      console.log("stop       : stop all sensors");
      console.log("throttle   : set motor");
      console.log("steering   : set servo angle");
      console.log("add point  : add waypoint for nav");
      console.log("enablenav  : enable navigation");
      console.log("disablenav : disable navigation");
      console.log("takepicture: take a picture");
      console.log("____________________________________");
      console.log("to exit press ctrl+c or exit or quit");

      this.complete();
  };

  this.update = function() {
    console.log('updating all sensors');
    this.kernel.update();
    for (kv in this.kernel.data) {
      console.log(kv + " = " + this.kernel.data[kv]);
    }

    this.complete();
  };

  this.config = function() {
    for (kv in this.kernel.config) {
      console.log(kv + " = " + this.kernel.config[kv]);
    }

    this.complete();
  };

  this.startComponents = function () {
    this.kernel.startComponents();

    this.complete();
  };

  this.stopComponents = function () {
    this.kernel.stopComponents();

    this.complete();
  };

  this.throttle = function (level) {
    this.ask("Throttle level -100 to 100: ",function (data){
      if (this.kernel.components.driver) {
        this.kernel.components.driver.setThrottle(data);
      } else {
        console.log("driver is not initialized yet");
      }

      this.complete();
    }.bind(this));
  };

  this.steering = function (angle) {
    this.ask("Steering angle -100 to 100: ",function (data){
      if (this.kernel.components.driver) {
        this.kernel.components.driver.setSteering(data);
      } else {
        console.log("driver is not initialized yet");
      }

      this.complete();
    }.bind(this));
  };

  this.halt = function () {
    this.components.driver.halt();

    this.complete();
  };

  this.exit = function () {
    console.log('exiting program');
    process.exit();
  };

  this.ask = function (question, callback) {
   this.stdout.write(question + ": ");
   this.stdin.once('data', function(data) {
     data = data.toString().trim();
     callback(data);
   });
 };

}

module.exports = Command;
